apis/stacks/v1alpha1/types.go

```diff
@@ -85,22 +86,54 @@ type StackInstallStatus struct {
 	StackRecord *corev1.ObjectReference `json:"stackRecord,omitempty"`
 }
 
-// Image returns the fully qualified image name for the StackInstallSpec.
-// based on the fully qualified image name format of hostname[:port]/username/reponame[:tag]
+// Image returns the Package prefixed with a source (if available).
+// If the package format is not understood it is returned unmodified to be
+// handled by Kubernetes.
 func (spec StackInstallSpec) Image() string {
-	if spec.Source == "" {
-		// there is no package source, simply return the package name
+	image, err := spec.ImageWithSource(spec.Package)
+	if err != nil {
 		return spec.Package
 	}
+	return image
+}
+
+// ImageWithSource applies a source to a container image URI only if the image
+// does not appear to contain a source. Source is some combination of scheme,
+// host, port, and prefix where host is required.
+func (spec StackInstallSpec) ImageWithSource(image string) (string, error) {
+	// no alternate source to substitute
+	if len(spec.Source) == 0 {
+		return image, nil
+	}
+
+	// prepend source to image when it does not have a source
+	named, err := reference.ParseNormalizedNamed(image)
+
+	if err != nil {
+		return "", err
+	}
+
+	// reference.Domain returns docker.io when no domain is found. If the image
+	// didn't explicitly start with docker.io, ignore it. In these cases, we
+	// want to apply the StackInstall source.
+	domain := reference.Domain(named)
+	if strings.Index(image, domain) != 0 {
+		return strings.Trim(spec.Source, "/") + "/" + image, nil
+	}
 
-	return fmt.Sprintf("%s/%s", spec.Source, spec.Package)
+	// image contained a source
+	return image, nil
 }
 
-// Image gets the Spec.Image of the StackInstall
-func (si *StackInstall) Image() string { return si.Spec.Image() }
+// GetSpec gets the Spec of the StackInstall
+func (si *StackInstall) GetSpec() *StackInstallSpec {
```

Comment:
does this need to return a pointer type? when it's a pointer it makes callers want/need to implement nil checks so make sure they don't have a nil pointer, which is extra work that could hopefully be avoided

Reply:
It can be avoid here. The only case that the pointer is useful now is for tests that provide StackInstall modifiers which I would like to add in #1281 (or could bring here).

The alternative is that I add `SetSource`, `SetImagePullPolicy`, `SetImagePullSecrets` which I am fine with adding.  `Get*` might remove my need for `GetSpec`.

Reply:
I've pushed this approach in the latest.

Reply:
Same change (SetSource, GetPackage) is now here in ef9be762. GetSpec is no longer introduced.